#1.1

SemekhinV opened the pull request:
add general structure

SemekhinV committed:
#1.1

add general structure

SemekhinV committed:
#1.2

added tests

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dao/films/FilmDaoImpl.java:
```diff
@@ -0,0 +1,251 @@
+
+        jdbcTemplate.batchUpdate(
+
+                "INSERT INTO film_genre (film_id, genre_id) VALUES (?,?)",
```
Здесь и в других методах тоже. 

Это не очень удачное решение-прямой доступ к чужой таблице. По идее, данный класс должен работать только и исключительно со своей таблицей. Лучше было бы выделить методы изменения таблицы "Фильмы-Жанры" в отдельный класс.

модификация данных в таблице должны происходить исключительно и только в соответствующем классе. 
Принцип такой: одна таблица- один класс для добавления, удаления, изменения данных в этой таблице.

SemekhinV replied:
Большое спасибо за совет, сделал, как и просили. Но у меня вопрос, является ли допустимой практикой вызов метода UserLikesDaoImpl в методе addLike класса UserDaoImpl или нужно было делать как-то по другому?

avfyodorov replied:
Вообще говоря, нет, это нехорошо. Репозитории (DAO) Ничего друг про друга знать не должны. Репозиторий выполняет эскуэль команды для своей таблицы и,  если надо,  возвращает результаты в сервис. 
Если же репозиторий вызывает другие репозитории, то есть фактически зависит от них, то это серьёзный повод пересмотреть архитектуру программы. 
Вся бизнес логика находится в сервисах. Сервис может обращаться к другим сервисам, а также получать данные из нескольких репозиториев. Репозиторий же настроен исключительно на свою таблицу.

SemekhinV replied:
Большое спасибо за совет и быструю проверку, вывел все на уровень сервисов.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/entity/Genre.java:
```diff
@@ -0,0 +1,13 @@
+@AllArgsConstructor
+public class Genre {
+
+    int id;
```
Здесь и дальше. Хорошо бы сделать эти поля приватными. 
   Можно воспользоваться аннотацией
@FieldDefaults(level = AccessLevel.PRIVATE)

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dao/users/UserDaoImpl.java:
```diff
@@ -0,0 +1,129 @@
+    private final JdbcTemplate jdbcTemplate;
+
+    @Override
+    public Optional<User> addEntity(User entity) {
```
Честно говоря, меня сбил с толку суффикс этого метода (addEntity). Речь идёт о пользователе, правильно? Так почему не назвать этот и остальные методы, например:  …….User?

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/service/users/UserService.java:
```diff
@@ -0,0 +1,27 @@
+
+public interface UserService {
+    //В качестве названия методов невозможно использовать просто add, delete
+    //И т.д., потому что возникает конфликт с сервисными методами
```
Да, понятно, здесь вопросов нет.

avfyodorov reviewed (commented):
Добрый вечер, Василий!

Серьёзное замечание только одно: одна таблица базы данных - один класс для выполнения базовых операций с этой таблицей. 
Я не стал писать одинаковые комментарии в разных местах программы. Пожалуйста,  просмотрите ещё раз все классы.

SemekhinV committed:
#2.1

added new classes with functionality for accessing tables (FilmLikes, FilmGenre, UserFriends) and some more minor improvements

avfyodorov reviewed (commented):
Добрый вечер, Василий!

На ваш вопрос ответил здесь, на гитхабе, в комментариях,  посмотрите пожалуйста. 

Давайте сделаем так: чтобы вас не задерживать ещё на одну итерацию, работу я приму сейчас. Вы же внесёте оставшиеся  изменения в программу с учётом обсуждения.

SemekhinV committed:
#3.1

redesigned and divided the DAO architecture in a ratio of 1 table - 1 DAO
